server/src/internalClusterTest/java/org/elasticsearch/index/engine/CheckAbortedMergesIT.java

```diff
@@ -0,0 +1,364 @@
+            .findFirst()
+            .orElseThrow(() -> new AssertionError("Plugin not found"));
+
+        final boolean checkAbortedMerges = randomBoolean();
+        pluginA.blockMerges();
+
+        final var indexName = randomIdentifier();
+        createIndex(
+            indexName,
+            indexSettings(1, 0).put(CheckAbortedDuringMergePolicy.ENABLE_CHECK_ABORTED_DURING_MERGE.getKey(), checkAbortedMerges)
+                .put(IndexSettings.INDEX_REFRESH_INTERVAL_SETTING.getKey(), TimeValue.MINUS_ONE)
+                .build()
+        );
+
+        var indexServiceA = internalCluster().getInstance(IndicesService.class, nodeA).indexService(resolveIndex(indexName));
+        assertThat(indexServiceA.hasShard(0), equalTo(true));
+
+        indexDocs(indexName, 10);
+        flush(indexName);
+
+        while (true) {
+            indexDocs(indexName, 10);
+            flush(indexName);
+
+            var mergesStats = client().admin().indices().prepareStats(indexName).clear().setMerge(true).get();
+            if (mergesStats.getIndices().get(indexName).getPrimaries().getMerge().getCurrent() > 0) {
+                break;
+            }
+        }
+
+        var nodeB = internalCluster().startDataOnlyNode();
+        ensureStableCluster(3);
+
+        pluginA.waitForMergesBlocked();
+
+        ClusterRerouteUtils.reroute(client(), new MoveAllocationCommand(indexName, 0, nodeA, nodeB, ProjectId.DEFAULT));
+        ensureGreen(indexName);
+
+        var indexServiceB = internalCluster().getInstance(IndicesService.class, nodeB).indexService(resolveIndex(indexName));
+        assertBusy(() -> assertThat(indexServiceB.hasShard(0), equalTo(true)));
+        assertBusy(() -> assertThat(indexServiceA.hasShard(0), equalTo(false)));
+        if (randomBoolean()) {
+            forceMerge();
+        }
+
+        assertThat(pluginA.mergedDocsCount.get(), equalTo(0L));
+        assertThat(pluginA.mergedFieldsCount.get(), equalTo(0L));
+        assertThat(pluginA.checkIntegrityCount.get(), equalTo(0L));
+
+        pluginA.unblockMerges();
+
+        var mergeMetrics = internalCluster().getDataNodeInstances(MergeMetrics.class);
+        assertBusy(
+            () -> assertThat(
+                StreamSupport.stream(mergeMetrics.spliterator(), false)
+                    .mapToLong(m -> m.getQueuedMergeSizeInBytes() + m.getRunningMergeSizeInBytes())
+                    .sum(),
+                equalTo(0L)
+            )
+        );
+
+        assertBusy(() -> {
+            if (checkAbortedMerges) {
+                assertThat(pluginA.mergedDocsCount.get(), equalTo(0L));
+                assertThat(pluginA.mergedFieldsCount.get(), equalTo(0L));
+                // Only the first integrity check is completed, the following ones should have been aborted
+                assertThat(pluginA.checkIntegrityCount.get(), equalTo(1L));
+            } else {
+                assertThat(pluginA.mergedDocsCount.get(), greaterThan(0L));
```

Comment:
Just for my understanding, Lucene would check the aborted flag likely at the beginning of the merge and that's why we see the same numbers here? wouldn't it make sense to let the merge make some progress so we actually see/check the effects of this PR code?